c8d: Implement Children by comparing diff ids

vvoland opened the pull request:
- Upstreams: https://github.com/rumpl/moby/pull/115

Implement Children method for containerd image store which makes the `ancestor` filter work for `docker ps`. 

Checking if image is a children of other image is implemented by comparing their rootfs diffids because containerd image store doesn't have a concept of image parentship like the graphdriver store. The child is expected to have more layers than the parent and should start with all parent layers.


**- How to verify it**

```bash
$ docker run -it alpine
/ # apk add fish
...
OK: 22 MiB in 24 packages

$ docker commit 8b1ea669e23e alpine-plus-fish
sha256:776b08f8a0cb27796cced75fcbe783bce0f2cbb0a91c38faadea44b519509c29

$ docker create alpine-plus-fish
c8e1caffdbbabdd3d67dbb9768d67a66ef98d26004eb4e10d7a880dd5da0870a

$ docker create busybox
7276114c22057a45964c93fcada3e23d2856e9021f1fdcc4959d410b8dce8eab

$ docker ps -a
CONTAINER ID   IMAGE              COMMAND     CREATED              STATUS                     PORTS     NAMES
7276114c2205   busybox            "sh"        About a minute ago   Created                              angry_kare
c8e1caffdbba   alpine-plus-fish   "/bin/sh"   2 minutes ago        Created                              brave_mcnulty
8b1ea669e23e   alpine             "/bin/sh"   2 minutes ago        Exited (0) 2 minutes ago             nifty_kilby

$ docker ps -a --filter ancestor=busybox
CONTAINER ID   IMAGE     COMMAND   CREATED         STATUS    PORTS     NAMES
7276114c2205   busybox   "sh"      2 minutes ago   Created             angry_kare
$ docker ps -a --filter ancestor=alpine-plus-fish
CONTAINER ID   IMAGE              COMMAND     CREATED         STATUS    PORTS     NAMES
c8e1caffdbba   alpine-plus-fish   "/bin/sh"   2 minutes ago   Created             brave_mcnulty
$ docker ps -a --filter ancestor=alpine
CONTAINER ID   IMAGE              COMMAND     CREATED         STATUS                     PORTS     NAMES
c8e1caffdbba   alpine-plus-fish   "/bin/sh"   2 minutes ago   Created                              brave_mcnulty
8b1ea669e23e   alpine             "/bin/sh"   3 minutes ago   Exited (0) 2 minutes ago             nifty_kilby

```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
c8d: Implement Children by comparing diff ids

Implement Children method for containerd image store which makes the
`ancestor` filter work for `docker ps`. Checking if image is a children
of other image is implemented by comparing their rootfs diffids because
containerd image store doesn't have a concept of image parentship like
the graphdriver store. The child is expected to have more layers than
the parent and should start with all parent layers.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on daemon/containerd/image_children.go:
```diff
@@ -0,0 +1,129 @@
+// rootfs of the given image ID, excluding images with exactly the same rootfs.
+// Called from list.go to filter containers.
+func (i *ImageService) Children(id image.ID) []image.ID {
+	ctx := context.TODO()
```
Should we add the context as a parameter?

vvoland replied:
It would be good to be able to cancel this operation because it can be heavy if there are a lot of images.
However, the interface function is used in other places in the graphdriver image store implementation which would also require context passing, which are not related to this particular implementation.
So I'd leave it for a follow up (handling context in bunch of other graphdriver image store code or refactoring the interface to make it possible to pass context only to this one).

vvoland replied:
Hmm actually it might be simpler. Let me try.

thaJeztah replied:
I know for (some of) the graph driver implementations we discussed that we are ok with only adding the context, but "ignoring it" / "not using it" if we can't (so only changing the interface, and add a `_ context.Context` where applicable)

vvoland replied:
Updated - I initially misread `Children` method from `image.Store` as the one from `ImageService` 😄 

https://github.com/moby/moby/blob/58c027ac8ba19a3fa339c65274ea6704ccbec977/image/store.go#L18-L27

vvoland committed:
ImageService: Pass ctx to Children

This only makes the containerd ImageService implementation respect
context cancellation though.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/containerd/image_children.go:
```diff
@@ -0,0 +1,127 @@
+		return []image.ID{}
+	}
+
+	is := i.client.ImageService()
```
nit; `is` looks to be only used at line 46, so perhaps we don't need the variable (or if we want to have a variable, move it closer to where it's used)

thaJeztah reviewed (approved):
LGTM, thanks!
